evals/completion_fns/openai.py

```diff
@@ -123,8 +124,12 @@ def __call__(
 
         openai_create_prompt: OpenAICreatePrompt = prompt.to_formatted_prompt()
 
+        client_id = (self.api_key, self.api_base)
+        if client_id not in CLIENTS:
+            CLIENTS[client_id] = OpenAI(api_key=self.api_key, base_url=self.api_base)
+
```

Comment:
Is there a need to have more than one client? Since the api key and base url are set in the ctor, seems like just caching the client in `self.client` would suffice.

Reply:
More than one client: I think it's possible to have two. One for completion_fn and one for metric (if model_graded). With `NestedSolver` it's possible to get more as well.

`self.client`: I believe multi-threading breaks if `client` is added to `self`, hence the ugly global variable.